crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -0,0 +1,327 @@
+pub struct Empty;
+
+/// Publish state after collecting nix metadata
+///
+/// Json value (ideally a [flox_types::catalog::CatalogEntry],
+/// but that's currently broken on account of some flakerefs)
+#[derive(Debug, Deref, DerefMut)]
+pub struct NixAnalysis(Value);
+
+/// State for the publish algorihm
+///
+/// Transitions through typestates to ensure we don't invoke invalid operations
+pub struct Publish<'flox, State> {
+    flox: &'flox Flox,
+    /// The published _upstream_ source
+    publish_ref: PublishRef,
+    /// The published attrpath
+    /// Should be fully resolved to avoid ambiguity
+    attr_path: AttrPath,
+    stability: Stability,
+    analysis: State,
+}
+
+impl<'flox> Publish<'flox, Empty> {
+    pub fn new(
+        flox: &'flox Flox,
+        publish_ref: PublishRef,
+        attr_path: AttrPath,
+        stability: Stability,
+    ) -> Publish<'flox, Empty> {
+        Self {
+            flox,
+            publish_ref,
+            attr_path,
+            stability,
+            analysis: Empty,
+        }
+    }
+
+    /// run analysis on the package and switch to next state
+    ///
+    /// It uses an analyzer flake to extract eval metadata of the derivation.
+    /// The analyzer applies a function to all packages in a `target` flake
+    /// and provides the result under `#analysis.eval.<full attrpath of the package>`.
+    ///
+    /// We evalaute this analysis as json, to which we add
+    /// * source urls for reproducibility
+    /// * the nixpkgs stability being used to create the package
+    pub async fn analyze(self) -> PublishResult<Publish<'flox, NixAnalysis>> {
+        let mut drv_metadata_json = self.get_drv_metadata().await?;
+        let flake_metadata = self.get_flake_metadata().await?;
+
+        // DEVIATION FROM BASH: using `locked` here instead of `resolved`
+        //                      this is used to reproduce the package,
+        //                      but is essentially redundant because of the `source.locked`
+        drv_metadata_json["element"]["url"] = json!(flake_metadata.locked.to_string());
+        drv_metadata_json["source"] = json!({
+            "locked": flake_metadata.locked,
+            "original": flake_metadata.original,
+            "remote": flake_metadata.original,
+        });
+        drv_metadata_json["eval"]["stability"] = json!(self.stability);
+
+        Ok(Publish {
+            flox: self.flox,
+            publish_ref: self.publish_ref,
+            attr_path: self.attr_path,
+            stability: self.stability,
+            analysis: NixAnalysis(drv_metadata_json),
+        })
+    }
+
+    /// extract metadata of the published derivation using the analyzer flake
+    async fn get_drv_metadata(&self) -> PublishResult<Value> {
+        let nix: NixCommandLine = self.flox.nix(Default::default());
+
+        let analysis_attr_path = {
+            let mut attrpath = AttrPath::try_from(["", "analysis", "eval"]).unwrap();
```

Comment:
Seems like this logic around `""` belongs in AttrPath

Reply:
AttrPath is a runix module and runix is supposed to be nix compliant.
I meant to bring support for absolute attrpaths into upstream but i wouldn't want to add it to runix just yet